src/server/src/grpc/mod.rs

```diff
@@ -294,26 +310,54 @@ impl Builder {
         self.query_dedup_config = Some(config);
         self
     }
+
+    // Compaction runner is an optional field for building [RpcServices].
+    pub fn compaction_runner(mut self, runner: Option<CompactionRunnerRef>) -> Self {
+        self.compaction_runner = runner;
+        self
+    }
 }
 
 impl Builder {
     pub fn build(self) -> Result<RpcServices> {
         let auth = self.auth.context(MissingAuth)?;
         let runtimes = self.runtimes.context(MissingRuntimes)?;
         let instance = self.instance.context(MissingInstance)?;
-        let opened_wals = self.opened_wals.context(MissingWals)?;
         let proxy = self.proxy.context(MissingProxy)?;
         let hotspot_recorder = self.hotspot_recorder.context(MissingHotspotRecorder)?;
-
-        let meta_rpc_server = self.cluster.map(|v| {
-            let builder = meta_event_service::Builder {
-                cluster: v,
-                instance: instance.clone(),
-                runtime: runtimes.meta_runtime.clone(),
-                opened_wals,
-            };
-            MetaEventServiceServer::new(builder.build())
-        });
+        let mut meta_rpc_server: Option<MetaEventServiceServer<MetaServiceImpl>> = None;
+        let mut compaction_rpc_server: Option<CompactionServiceServer<CompactionServiceImpl>> =
+            None;
+
+        self.cluster
+            .map(|v| {
+                let result: Result<()> = (|| {
+                    // Support meta rpc service.
+                    let opened_wals = self.opened_wals.context(MissingWals)?;
+                    let builder = meta_event_service::Builder {
+                        cluster: v.clone(),
+                        instance: instance.clone(),
+                        runtime: runtimes.meta_runtime.clone(),
+                        opened_wals,
+                    };
+                    meta_rpc_server = Some(MetaEventServiceServer::new(builder.build()));
+
+                    // Support remote compaction rpc service.
+                    let compaction_runner =
+                        self.compaction_runner.context(MissingCompactionRunner)?;
+                    let builder = compaction_service::Builder {
```

Comment:
Maybe the compaction service is not something so related to the `Cluster`.
`Cluster` is used by the physical node to operate the `virtual node`(aka `shard`).

Currently, I think we can jsut always open the compaction service in all horaedb node?

Reply:
Right, the `Cluster` is unnecessary, i should remove it. 

And the current implementation already always open the compaction service in all horaedb nodes.